CHANGES.md

```diff
@@ -43,6 +44,15 @@ acknowledgements of a `receive` related command. The user is free to inspect the
 to handle it - including even throwing an exception if OK status (if this makes sense for whatever reason). The default implementation checks for ERROR code and throws
 an exception if so.
 
+### OpenSSH protocol extension: strict key exchange
+
+[GH-445](https://github.com/apache/mina-sshd/issues/445) implements an extension to the SSH protocol introduced
+in OpenSSH 9.6. This ["strict key exchange" extension](https://github.com/openssh/openssh-portable/blob/master/PROTOCOL)
+hardens the SSH key exchange against the ["Terrapin attack"](https://www.terrapin-attack.com/). The extension
+is active if both parties announce their support for it at the start of the initial key exchange. If only one
```

Comment:
As mentioned, I think it will help admins if they can set a flag to disconnect connections if they do not negotiate strict-kex. This makes is much easier to re-enable the vulnerable (but modern) ciphers. Maybe it’s an option for later, but why not offer it from the get-go. If not offered maybe a sample listener who does it would be good (validating the info is exposed to make that call by the app).

nb: I am not saying make sending of the offer configurable, it’s perfectly fine with me to do that unconditional (even though some other impls offer this option).

Reply:
A "RequireStrictKex" config for `~/.ssh/config` (default "no") may make sense. The proposal has come up before, for instance on the SSH mailing list.

It could be added in a follow-up change. My main motivation for this PR was to have a minimal change with sufficient tests so that we can be confident it works, and that it works also when talking to peers that don't have strict kex at all.

A problem with custom configs is that they cause OpenSSH to complain and fail unless the user also adds an `IgnoreUnknown RequireStrictKex` to the config file. But for this option you probably don't want to ignore it. I didn't see such a config option in OpenSSH. So I think we should add such an option only if OpenSSH has it.

Reply:
Ok, I am not sure who would actually use Mina-sshd for such things, but yes a common config option would be great. I didn’t receive any reaction to my OpenSSH-dev question, though.

Jsch has a enable (aka offer/support) and require option.

Reply:
I'm aware of the option in JSch 0.2.15. A `CoreModuleProporties` property for requiring strict kex and failing the connection if strict kex is not negotiated in `negotiate()` could be considered, and could easily be added in a follow-up commit. Or we could add it to the negotiation result in a follow-up commit; then a `SessionListener.sessionNegotiationEnd()` could be used to close the connection by throwing an exception.

To be really useful one would need the `~/.ssh/config` option, though. And as I wrote, I believe that should be done only if OpenSSH gets it, too.

In any case, IMO this is something for follow-up changes.

Reply:
About the "offer" option: see my comment about a "kill switch" below. Users can already switch strict kex off if they absolutely want to do so. But it should require a bit more conscious thought and effort than just flipping a boolean flag.